upbc/protoc-gen-upb.cc

```diff
@@ -512,11 +541,19 @@ void GenerateMessageInHeader(const protobuf::Descriptor* message, Output& output
           GetSizeInit(layout.GetOneofCaseOffset(field->real_containing_oneof())),
           field->number(), FieldDefault(field));
     } else {
-      output(
-          "UPB_INLINE $0 $1_$2(const $1 *msg) { "
-          "return *UPB_PTR_AT(msg, $3, $0); }\n",
-          CTypeConst(field), msg_name, field->name(),
-          GetSizeInit(layout.GetFieldOffset(field)));
+      if (HasNonZeroDefault(field)) {
+        output(
+            "UPB_INLINE $0 $1_$2(const $1 *msg) { "
+            "return $1_has_$2(msg) ? *UPB_PTR_AT(msg, $3, $0) : $4; }\n",
```

Comment:
given that we are in C++ here

R"cc(
UPB_INLINE $0 $1_$2(const  $1 *msg) {
  return ...
})cc"

is a bit easier to scan (and I think clang-format will even format the island grammar for you)

Reply:
Done -- the clang-format part is very nifty. I changed the output routine to strip the leading indentation so the generated code doesn't drift right.

Reply:
I hope to one day refactor the bulk of our code generator to use this technique, but I cannot justify the cost in any way :)